### PR TITLE
Auto select https if port is 443 and not protocol is detected

### LINKED ISSUE #37490: Magento 2.4.6 AWS Opensearch 1.2. No Alive nodes found.
### Preconditions and environment

- Magento 2.4.6
- AWS OpenSearch Domain 1.2

### Steps to reproduce

1. Clone version 2.4.6
2. Configure opensearch in env.php
3. Run `php bin/magento setup:install` or `php bin/magento setup:upgrade`
4. Then the error happens when it checks Opensearch

### Expected result

Magento should be able to configure the domain and use it

### Actual result

Error code displayed:
image

### Additional information

My env.php Opensearch configuration:
s

Same machine using curl to check cluster health, everything seems ok:
image

Opensearch general information:
image

Opensearch security config:
image

Opensearch access policy:
```
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "AWS": "*"
      },
      "Action": "es:*",
      "Resource": "arn:aws:es:eu-west-3:<MY_ACC>:domain/magento-opensearch-dev/*"
    }
  ]
}
```

My Cluster Config:
image
image

AWS Setup:
```
VPC
|- Opensearch in private subnet
|- EC2 instance with magento in private subnet
```

AWS Support when I asked for help regarding this:
```
Warm Greetings from AWS Premium Support. Thank you for reaching out to me with your concern. My name is Risha and I will be assisting you with the case today.

From the case notes, I understand that you are getting the error "could not validate a connection in Opensearch. No alive nodes found in your cluster". 

Kindly note that I checked the cluster heath and it is "Green" and it have never been yellow or red. I observed the last 3 days Clouldwatch metrics. 

Please find below the output of "GET /_cluster/health". 
-----------------
{
    "cluster_name": "<MY_ACC>:magento-opensearch-dev",
    "status": "green",
    "timed_out": false,
    "number_of_nodes": 1,
    "number_of_data_nodes": 1,
    "discovered_master": true,
    "active_primary_shards": 1,
    "active_shards": 1,
    "relocating_shards": 0,
    "initializing_shards": 0,
    "unassigned_shards": 0,
    "delayed_unassigned_shards": 0,
    "number_of_pending_tasks": 0,
    "number_of_in_flight_fetch": 0,
    "task_max_waiting_in_queue_millis": 0,
    "active_shards_percent_as_number": 100
}
----------------

Further, I checked CPUUtilization and JVMMemoryPressure and I confirm that all the metrics are within threshold.

I also observed the nodes metric and I understand that you had 2 nodes previously, then you enabled master nodes at around 2023-05-09 20:50 UTC for 20 mins then at around 2023-05-10 10:05 UTC you have decreased one more node. At present, you have only 1 node in the cluster. 

At present, the cluster seems in a good state as of now and their is no such abnormal activities in the cluster. I also observed that your cluster is within a VPC, therefore it is possible this error would be happened on your client side or client machine due to the network limit/throttle from your client to our opensearch cluster. Since I checked your Opensearch cluster, there is no abnormal activity in the last 3 day and the cluster is in active state.

I kindly suggest you to check your client side. I hope the above assistance is helpful to you. However, if you have any other query/concern, kindly write back to me and I would be more than happy to assist you further. 

Thank you and have a nice day ahead. 
Stay safe!

We value your feedback. Please share your experience by rating this and other correspondences in the AWS Support Center. You can rate a correspondence by selecting the stars in the top right corner of the correspondence.
```

Already tried having them both on the same subnet but didn't change anything

### Release note

_No response_

### Triage and priority

- [X] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by DEADSEC-SECURITY**
Summary:
Fixes solution 1 of Issue 37490 by defaulting to protocol https if no protocol is given in the hostname and the port given is 443

Issues:
https://github.com/magento/magento2/issues/37490

Testing:
Tested on 2.4-develop on AWS EC2 instance with AWS Opensearch domain running version 1.3

### Fixed Issues (if relevant)

1. Partially fixes magento/magento2#37490

### Manual testing scenarios (*)

1. Create a fresh instance, in opensearch params use port 443 and don't provide a protocol in hostname, the resulting hostname should have protocol https instead of the original default to http

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [x] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by DEADSEC-SECURITY** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
Hello @DEADSEC-SECURITY,

Thanks for the contribution!

But in order to proceed further please sign the Adobe CLA. Please check the build failures for more details.

Thanks

**Comment by DEADSEC-SECURITY** [comment]
> Hello @DEADSEC-SECURITY,
> 
> Thanks for the contribution!
> 
> But in order to proceed further please sign the Adobe CLA. Please check the build failures for more details.
> 
> Thanks

I already signed the CLA.

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by hostep** [comment]
> I already signed the CLA.

There is a long standing bug with Adobe's CLA check here on github. To fix it, you need to close the PR, and a couple of seconds later re-open it again. From then on, it should work. This is only needed for your first PR to this repo here. Kind of annoying, but it's an easy fix 🙂

**Comment by DEADSEC-SECURITY** [comment]
that worked @hostep

File: app/code/Magento/OpenSearch/Model/SearchClient.php [hunk 1]
```diff
@@ -142,6 +142,11 @@ private function buildOSConfig(array $options = []): array
         // @codingStandardsIgnoreEnd
         if (!$protocol) {
             $protocol = 'http';
+
+            // If port is 443 then use https
+            if (!empty($options['port']) && $options['port'] == '443') {
```
Comment (engcom-Hotel) [inline app/code/Magento/OpenSearch/Model/SearchClient.php]: Suggest you to implement strict check here like below:

`$options['port'] === '443'`

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]
Hello @DEADSEC-SECURITY,

Please have a look on my suggestions below and also please cover the changes with some automated tests.

Thanks

**Comment by DEADSEC-SECURITY** [comment]
> Hello @DEADSEC-SECURITY,
> 
> Please have a look on my suggestions below and also please cover the changes with some automated tests.
> 
> Thanks

Hi @engcom-Hotel I'm new to the Magento codebase and PHP is not the main language so if you could point me out to docs on how to creates tests I would appreciate.

**Pull request closed without merging**